プラクティス「wc コマンドを作る」の課題提出

nekonote-civ opened the pull request:
# 概要
プラクティス「wc コマンドを作る」の課題提出

# 詳細
プラクティス課題を作成し、コマンド実行まで正常に動作したことをローカルにて確認しました。
お手数おかけしますがレビューをよろしくお願いします！

nekonote-civ committed:
初期設定用の各ファイルを作成

nekonote-civ committed:
[-l, -w, -c] 各オプションの実装

-l: 行数(改行コードをベースに算出)
-w: 単語数(半角スペース、タブ、改行コード区切り)
-c: バイト数(bytesize メソッドを使用)

nekonote-civ committed:
オプションなしの場合の出力幅を wc コマンドと合わせた

nekonote-civ committed:
標準入力を含めた各パターン別の出力処理を作成

nekonote-civ committed:
重複処理をメソッドに切り出し、変数名等を状況に即したものへ変更した

nekonote-civ committed:
ファイルを複数指定した場合の出力処理を作成

nekonote-civ committed:
重複処理のメソッド切り出しとリネームを実施し、伝わりづらい箇所へコメントを追加

nekonote-civ committed:
コメントの追加やリネーム、短い表記への変更、メソッドへの切り出しを実施

nekonote-civ committed:
最終行の改行を追加

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

nekonote-civ committed:
メイン処理をトップに記述するように変更

nekonote-civ committed:
各カウントを hash で扱うように変更し、パイプのパターンのメソッドを上から読めるように並び替えた

nekonote-civ committed:
ファイル指定の場合の出力方法をシンプルにまとめ、分岐数を減らした

nekonote-civ committed:
values.map を使用するように修正

nekonote-civ committed:
不要なファイルを削除

nekonote-civ commented:
> 動画でレビューしたので別途リンクを送ります〜。

レビューありがとうございます！
修正対応しましたので、お手数をおかけしますが再レビューを宜しくお願い致します。

JunichiIto reviewed (commented):
今回も動画でレビューしたので別途リンクを送ります〜。

nekonote-civ committed:
不要なコメントを削除

nekonote-civ committed:
一度 max してから 文字列幅を取得するように修正

nekonote-civ committed:
二人三脚になってしまっていた部分を排除し、Hash の key, value を活用するように修正

nekonote-civ commented:
ご指摘の内容を再度修正しましたので、レビューを宜しくお願い致します。

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,95 @@
+
+  return if single_file?
+
+  total_counts = sum_counts(counts_array)
```
37行目でも`sum_counts`を呼んでるので同じ処理を2回呼ぶのは非効率だなーと思いました。

nekonote-civ replied:
確かに2回利用しているので無駄が発生していますね。
呼び出し箇所を1箇所に変更しました。

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,95 @@
+  list << counts[:line].to_s.rjust(length) if params[:l]
+  list << counts[:word].to_s.rjust(length) if params[:w]
+  list << counts[:size].to_s.rjust(length) if params[:c]
+  list << counts[:file_name] if counts[:file_name] && !counts[:file_name].empty?
```
```suggestion
  list << counts[:file_name] if !counts[:file_name].to_s.empty?
```

とするとシンプルになりそうです

nekonote-civ replied:
こちらご指摘のとおりに修正しました！
(ただし、 `if !counts...` ですと rubocop に怒られてしまうので、`unless` を使用しました)

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,95 @@
+
+def join_counts(counts, length, params)
+  list = []
+  list << counts[:line].to_s.rjust(length) if params[:l]
```
このロジックだと`wc.rb hoge.txt`のように何もオプションを指定しなかったときにファイル名しか表示されない、という問題が起きませんか？

nekonote-civ replied:
こちらは完全にミスです！
コミットが漏れてしまっていましたので追加しました。

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,95 @@
+  return if single_file?
+
+  total_counts = sum_counts(counts_array)
+  puts "#{join_counts(total_counts, length, params)} 合計"
```
`sum_counts`の中で:file_nameに"合計"をセットするのもありかなーと思いました。
そうするとjoin_countsのロジックを再利用できますし。

nekonote-civ replied:
確かにそのとおりですね、全く思いつきませんでした！
`sum_counts` にてセットするように変更し、呼び出し先の箇所で `max` を使い文字列長を判定する処理など、影響する部分を一部修正しました。

JunichiIto reviewed (commented):
コメントしました！

nekonote-civ committed:
漏れてしまっていた条件分岐を追加

nekonote-civ committed:
重複している処理を1回のみにした

nekonote-civ committed:
ファイル名の判定処理を簡略化

nekonote-civ committed:
sum_counts で file_name を持つように変更

nekonote-civ commented:
ご指摘事項を修正しましたので、再レビューを宜しくお願い致します🙇‍♂️

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,96 @@
+  unless single_option?(params) && single_file?
+    length = total_counts.map do |key, value|
+      key != :file_name ? value : 0
+    end.max.to_s.length
```
```suggestion
    length = total_counts.values_at(:line, :word, :size).max.to_s.length
```

と書くこともできそうです

nekonote-civ replied:
「キーを絞り込む」という観点が抜けていたので、今後はこういった活用が出来ることを意識してみます！
また、処理部が1行になったことで Rubocop の警告が出てしまうので、`unless` を後部に移動しました。

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,96 @@
+  list << counts[:line].to_s.rjust(length) if params[:l] || params.empty?
+  list << counts[:word].to_s.rjust(length) if params[:w] || params.empty?
+  list << counts[:size].to_s.rjust(length) if params[:c] || params.empty?
+  list << counts[:file_name] unless counts[:file_name].to_s.empty?
```
今の実装だと`to_s`をなくしてしまっても結果は変わらない気がするのですがどうでしょうか？

nekonote-civ replied:
`:file_name` に関しては確かに「空文字」or「ファイル名」しか入ってこない様になったので、むだな `to_s` を削除しました。
それ以外は Integer で渡されてくるため、引き続き `to_s` を使用するようにしています。

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,96 @@
+  list = []
+  list << counts[:line].to_s.rjust(length) if params[:l] || params.empty?
+  list << counts[:word].to_s.rjust(length) if params[:w] || params.empty?
+  list << counts[:size].to_s.rjust(length) if params[:c] || params.empty?
```
```suggestion
  list = %i(line word size).map do |key|
    counts[key].to_s.rjust(length) if params[key] || params.empty?
  end
```

paramsのkeyもcountsと同じにするとこんな書き方ができそうです

nekonote-civ replied:
`:line ~ :size` に関しては「シンボル」のみが変わるだけでしたので、ご指摘の通りに `map` で処理をするようにしました。

また、今までは「条件に当てはまった場合のみ list へ格納される」仕組みでしたが、上記の場合は `wc.rb -l` といった指定の場合には `nil` が含まれた配列が作成されるため、`reject` で弾くようにしました。

その他、`params[key]` には `-l` のようなパラメータが入る状態でしたので、`option_params` メソッドの以下の部分を変更して `:line` などの後に使用する名称に合わせました。

```ruby
  opt.on('-l') { |v| params[:line] = v }
  opt.on('-w') { |v| params[:word] = v }
  opt.on('-c') { |v| params[:size] = v }
```

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,96 @@
+def sum_counts(counts_array)
+  total_counts = { line: 0, word: 0, size: 0, file_name: '合計' }
+  counts_array.each do |counts|
+    counts.each do |key, value|
+      total_counts[key] += value if key != :file_name
+    end
```
```suggestion
    counts.except(:file_name).each do |key, value|
      total_counts[key] += value
    end
```

exceptを使うとコードの意図がより明確になるかもしれません。

https://docs.ruby-lang.org/ja/latest/method/Hash/i/except.html

nekonote-civ replied:
`except` メソッドというものがあるのですね！
ご指摘の通り、`if key != :file_name` と判定を入れずに `except` にて Hash のデータを絞り込む様に変更しました。

JunichiIto reviewed (commented):
コメントしました〜。

nekonote-civ committed:
join_counts の結合処理を map で簡略化した

map で処理する為に option_params の代入時のシンボルを変更した
counts[:file_name] は常に文字列であるため to_s を削除した

nekonote-civ committed:
except で対象外のキーを明示的にした

nekonote-civ committed:
values_at でキーを指定して Hash の最大値を判定するようにした

nekonote-civ commented:
再度修正しましたので、お手数をおかけしますがレビューを宜しくお願い致します🙇‍♂️

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -67,11 +63,10 @@ def single_file?
-  list << counts[:word].to_s.rjust(length) if params[:w] || params.empty?
-  list << counts[:size].to_s.rjust(length) if params[:c] || params.empty?
-  list << counts[:file_name] unless counts[:file_name].to_s.empty?
+  list = %i[line word size].map do |key|
+    counts[key].to_s.rjust(length) if params[key] || params.empty?
+  end.reject(&:nil?)
```
```suggestion
  list = %i[line word size].filter_map do |key|
    counts[key].to_s.rjust(length) if params[key] || params.empty?
  end
```

で同じことができるはずですー

nekonote-civ replied:
なるほど、`reject`よりも更に直感的ですね！

このあたりのリファレンスもしっかり読み直しておきます！
`https://docs.ruby-lang.org/ja/latest/class/Enumerable.html`

JunichiIto commented on 05.wc/wc.rb:
```diff
@@ -34,11 +34,7 @@ def print_type_file(params)
-      key != :file_name ? value : 0
-    end.max.to_s.length
-  end
+  length = total_counts.values_at(:line, :word, :size).max.to_s.length unless single_option?(params) && single_file?
```
ひとりごと）Rubocopの指摘ということなので、このままで良いですが、個人的にはとても見にくいと思うので僕だったら

```ruby
unless ...
  ...
end
```

の形式で書くな〜と思いました。

nekonote-civ replied:
そうですね、私もちょっと見づらいな～と思いました。
今後こういった場合には「警告は出るが可読性が悪いためこうしています」と沿えて複数行で書くなども考えてみます！

JunichiIto reviewed (approved):
1点コメントしましたが、大きな問題ではないのでこれで合格とします！ 🎉

nekonote-civ commented:
ありがとうございます！
それではOKということでPRはクローズとさせていただきます！
